### PR TITLE
Disallow custom units

### PR DESCRIPTION
**Pull request by jl-wynen**
Affects LLNL/Units custom units like CXCUN[...], CXUN[...], and equation units.

### CODE REVIEW

File: lib/units/test/unit_test.cpp [hunk 1]
```diff
@@ -31,6 +31,15 @@ TEST(UnitTest, construct_bad_string) {
   EXPECT_THROW(Unit("abcde"), except::UnitError);
 }
 
+TEST(UnitTest, custom_unit_strings_get_rejected) {
+  // Custom (counting) units and equation units are rejected.
+  for (const auto &str :
+       {"CXUN[0]", "CXUN[51]", "CXUN{1023]", "CXCUN[0]", "CXCUN[15]",
```
Comment (SimonHeybrock) [inline lib/units/test/unit_test.cpp:37]: Is the `{1023]` intentionally using mixed bracket types?

File: lib/units/unit.cpp [hunk 2]
```diff
@@ -25,12 +25,29 @@ std::string map_unit_string(const std::string &unit) {
...
+    llnl::units::precise::count.base_units());
+inline constexpr auto radian_eq_type = llnl::units::precise::custom::eq_type(
+    llnl::units::precise::radian.base_units());
+
+bool is_special_unit(const llnl::units::precise_unit &unit) {
+  using namespace llnl::units::precise::custom;
+  const auto &base = unit.base_units();
+
+  // count and radian are flagged as equation units, so we need to catch those
+  // early.
```
Comment (SimonHeybrock) [inline lib/units/unit.cpp:39]: ... but they are not using the flags that we are not storing in the file, are they?

File: lib/units/test/unit_test.cpp (hunk 1 above)
Comment (jl-wynen) [reply lib/units/test/unit_test.cpp:37]: No, it's a typo

File: lib/units/unit.cpp (hunk 2 above)
Comment (jl-wynen) [reply lib/units/unit.cpp:39]: No:
```.cpp
const auto u = units::Unit("count").underlying();
std::cout << llnl::units::precise::custom::eq_type(u.base_units()) << '\n';  // 8
std::cout << u.is_equation() << '\n';  // 0
std::cout << u.is_per_unit() << '\n';  // 0
std::cout << u.has_e_flag() << '\n';  // 0
std::cout << u.has_i_flag() << '\n';  // 0
```
`eq_type` has special cases for counts and radians but the other four lines (which are the ones used for serialization) don't reflect this. Instead, `count` uses its own base.

File: lib/units/test/unit_test.cpp [hunk 3]
```diff
@@ -31,6 +31,15 @@ TEST(UnitTest, construct_bad_string) {
   EXPECT_THROW(Unit("abcde"), except::UnitError);
 }
 
+TEST(UnitTest, custom_unit_strings_get_rejected) {
+  // Custom (counting) units and equation units are rejected.
+  for (const auto &str : {"CXUN[0]", "CXUN[51]", "CXUN[1023]", "CXCUN[0]",
+                          "CXCUN[15]", "decibels", "ln(x)", "EQXUN[1]",
```
Comment (SimonHeybrock) [inline lib/units/test/unit_test.cpp:37]: How did `decibels` work previously? I suppose Scipp was completely broken with unit conversion and arithmetic ops?

File: lib/units/test/unit_test.cpp [hunk 4]
```diff
@@ -238,7 +247,7 @@ TEST(UnitParseTest, singular_plural) {
 TEST(UnitFormatTest, roundtrip_string) {
   for (const auto &s :
        {"m", "m/s", "meV", "pAh", "mAh", "ns", "counts", "counts^2",
-        "counts/meV", "1/counts", "counts/m", "Y", "M", "D"}) {
+        "counts/meV", "1/counts", "counts/m", "rad", "$", "Y", "M", "D"}) {
```
Comment (SimonHeybrock) [inline lib/units/test/unit_test.cpp:250]: Why is `$`?

Comment (jl-wynen) [reply lib/units/test/unit_test.cpp:250]: For the 'currency' base. LLNL/Units represents all currencies as `$`

File: lib/units/test/unit_test.cpp (hunk 3 above)
Comment (jl-wynen) [reply lib/units/test/unit_test.cpp:37]: I think conversions didn't work, e.g.
```.py
(1.2*sc.Unit('decibel')).to(unit='m decibel')
```
raises
```
UnitError: Conversion from `EQXUN[3]` to `0.001*EQXUN[3]` is not valid.
```
(it is an equation unit, FYI)

File: lib/units/test/unit_test.cpp (hunk 4 above)
Comment (SimonHeybrock) [reply lib/units/test/unit_test.cpp:250]: I didn't even realize there *was* a currency base :D

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**